### PR TITLE
Configurable forms (feedback, acquisition request etc)

### PR DESCRIPTION
**Pull request by samuli**
Add support for configurable forms. Forms are configured in FeedbackForms.json and handled by `/Feedback/<form-id>` controller action.  The configuration includes the current Feedback form and an example for acquisition reqest.

The following options are used in configuring a form:

- enabled: Is the form enabled?
- title: Form title
- recipient: Recipient name and email
- emailSubject: Email subject
- allowLocalOverride: Can the form be overridden using local configuration (default: true)
- onlyForLoggedUsers: Show form only for logged users (default: false)
- response: Response shown after successful form submit
- senderInfoRequired: Require sender name & email fields to be filled?
- help: Form help texts

- fields: Array of form elements with the following options:

- name: Element name
- label: Element label
- type: Element type (text, textarea, select, email, url)
- required: Require element to be filled?
- settings: Array of key-value pairs that are mapped to HTML element attributes
- For select elements one of:
    - options: Array of options
    - optionGroups: Array of group -> options entries that are mapped to HTML optgroups.

### CODE REVIEW

**Comment by demiankatz** [comment]
@samuli, thanks for this substantial contribution! I'll take a closer look and offer some comments after the VuFind Summit next week.

File: themes/bootstrap3/theme.config.php [hunk 1]
```diff
@@ -27,6 +27,18 @@
     'favicon' => 'vufind-favicon.ico',
     'helpers' => [
         'factories' => [
+            'Zend\Form\View\Helper\Form' => 'Zend\ServiceManager\Factory\InvokableFactory',
```
Comment (demiankatz) [inline themes/bootstrap3/theme.config.php]: Can you avoid this extra configuration by activating the `Zend\Form` module here: https://github.com/vufind-org/vufind/blob/master/config/application.config.php#L5 ?

File: composer.json [hunk 2]
```diff
@@ -70,7 +70,8 @@
         "zendframework/zendrest": "2.0.2",
         "zendframework/zendservice-amazon": "2.3.0",
         "zendframework/zendservice-recaptcha": "3.1.0",
-        "zf-commons/zfc-rbac": "2.6.3"
+        "zf-commons/zfc-rbac": "2.6.3",
+        "zendframework/zend-form": "^2.12"
```
Comment (demiankatz) [inline composer.json]: I would recommend pinning this to 2.12.0 for consistency with the way we handle other components. We've had problems in the past where even minor version changes have broken things, so I try to keep versions under tight control to prevent a `composer update` from having unanticipated side effects.

File: themes/bootstrap3/theme.config.php (hunk 1 above)
Comment (samuli) [reply themes/bootstrap3/theme.config.php]: Yes, much better this way.

**Comment by demiankatz** [comment]
@samuli, I'm trying to get this running on my test server, but it's failing with an error that the Zend\Form module cannot be initialized. Is this code working for you as-is on your end? I apologize if I've lead you astray with my module initialization suggestion -- but perhaps there's some other problem on my end.

I'm also wondering if there is some code missing from this PR, because when I review the code, I can't seem to figure out where the JSON configuration file is getting read in. Am I overlooking something, or is there something else that needs to be added here?

Thanks for your help -- please let me know if you need me to do more troubleshooting on my end!

**Comment by samuli** [comment]
Sorry, forgot to include Form.php...

composer.lock was outdated, which hopefully explains why Zend\Form didn't work.

**Comment by demiankatz** [comment]
Thanks, @samuli, that all makes sense! I'll try again in the near future!

**Comment by demiankatz** [comment]
@samuli, I can confirm that this is now running successfully on my test server. Thanks again! I'll try to find time for an in-depth review next week.

**Comment by demiankatz** [comment]
@samuli, now that I've had time to play with this in a little more detail, the next thing I'm thinking about is backward-compatibility. I think we could probably justify including this feature in release 5.1 if we implement it in a slightly more back-compatible manner. What if, by default, the default feedback form is set up to look just like the old hard-coded form (with just Name, Email and Comments), with the other forms disabled by default? Also, what if instead of removing the recipient_email/recipient_name/email_subject settings from config.ini, you allowed those to continue to act as defaults if not overridden through the JSON file? That would allow existing installations to continue to function exactly the same as before with no configuration changes, but it would still open up all the extra functionality that you have developed for those who need it.

I'm happy to help make these adjustments if that would be useful, though I notice that I do not have push access to your branch (I tried to update by merging the latest master but could not), so that may make it a little less convenient for me to directly contribute to this work.

Let me know how you would like to proceed; I'll delve deeper into code review after we've sorted out these strategic questions.

**Comment by samuli** [comment]
The Feedback form should now behave like in the current version: recipient info and email subject is read by default from config.ini and the form only includes the message and sender name/email fields. 

One small change is that sender fields are always positioned at the end of the form, before the submit button. If this is a problem, we could add a ocnfiguration option for the position (beginning vs. end)? Another way would be to treat sender name and email fields as other form elements, so that they need to be included in the form configuration 'fields' array.

I assigned collaborator rights for you, so you should be able to make changes to the PR.

**Comment by demiankatz** [comment]
@samuli, thanks, this looks great. I'll take a closer look very soon!

Regarding the order of the form fields, I have no strong objection to changing the order... but if the JSON file takes order into account, I imagine it should be easy to customize this if anyone has a strong feeling about it.

**Comment by demiankatz** [comment]
@samuli, thanks again for this -- it is working well for me. I have made a couple of very minor adjustments so that the errors provide more appropriate HTTP status codes and to align the new form text with the pre-existing form text. A few comments and questions, though:

1.) I notice that right now, if I set values in the [Feedback] section of config.ini, they override all of the equivalent settings in FeedbackForms.json, rather than acting as defaults when FeedbackForms.json provides nothing more specific. Is that intentional? It seems backwards to me.

2.) The biggest drawback to the JSON format is that it does not support comments. Did you consider a different format like YAML that might allow us to make the configuration more self-documenting? Failing that, do you think it's worth adding an "ExampleForm" disabled form at the top of the file that describes all of the fields in order to serve as documentation? Another option might be to simply add an element above the top "forms" called "documentationUrl" that points to a new wiki page. If you like that approach, I can set up a wiki page for you to edit.

3.) As currently implemented, the \VuFind\Form\Form class is not easily extensible. Did you consider implementing this as a service with a factory? Perhaps instead of passing the form ID to the constructor, you could have a setFormId method... this would allow you to pull the object once from the service locator without having to set up so many dependencies in the controller. If you think this might be a worthwhile approach but need help setting it up, let me know and I'll be happy to work on it.

Thanks!

**Comment by samuli** [comment]
1. Yes, this was intentional, but maybe not so well thought out... I'll change FeedbackForms.json to override config.ini.

**Comment by samuli** [comment]
@demiankatz Points 1-3 are now fixed.

I had to make a few modifications to YAMLReader in order to be able to bypass file cache so that I can first read the default configuration and possibly re-read with local overrides after this.

**Comment by demiankatz** [comment]
@samuli, thanks for this, and sorry for introducing that enabled bug during my review editing. :-) I didn't realize you were still actively working on this and hadn't gotten around to testing it before you fixed it!

File: config/vufind/FeedbackForms.yaml [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+forms:
+
+  # Main Feedback form
+
+  FeedbackSite:
+    allowLocalOverride: false
+    enabled: true
+    title: Send us your feedback!
+    recipient:
```
Comment (demiankatz) [inline config/vufind/FeedbackForms.yaml]: Any objection to commenting this section out by default to let the config.ini defaults kick in for backward compatibility?

File: config/vufind/FeedbackForms.yaml [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      - name: message
+        type: textarea
+        label: Comments
+        required: true
+        settings:
+
+  # Example for an acquisition request form
+
+  AcquisitionRequest:
+    enabled: true
```
Comment (demiankatz) [inline config/vufind/FeedbackForms.yaml]: Any objection to making this false by default?

File: languages/en.ini [hunk 5]
```diff
@@ -382,6 +382,9 @@ fav_list_delete_fail = "Sorry, an error has occurred. Your list was not deleted.
 Favorites = "Saved Items"
 Fee = "Fee"
 Feedback = "Feedback"
+feedback_email = "Email"
+feedback_email_subject = "VuFind feedback"
```
Comment (demiankatz) [inline languages/en.ini]: Do we really need to translate the feedback email subject? The email is coming to the maintainer of VuFind, not the end user. Thus, it seems that you would want a uniform subject heading that is meaningful to the maintainer and that could be filtered on, not a subject that is translated into the end user's language.

**Comment by samuli** [comment]
@demiankatz No problem. I'm done for now, so feel free to continue editing.

File: module/VuFind/config/module.config.php [hunk 6]
```diff
@@ -677,6 +692,7 @@
 
 // Define dynamic routes -- controller => [route name => action]
 $dynamicRoutes = [
+    'Feedback' => ['form' => 'Form/[:id]'],
```
Comment (demiankatz) [inline module/VuFind/config/module.config.php]: Is this route used? I haven't had a chance to do testing, but just curious if we need both the feedback-form route above and this dynamic route, or if one is enough.

File: module/VuFind/src/VuFind/FormFactory.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        if (!empty($options)) {
+            throw new \Exception('Unexpected options sent to factory.');
+        }
+
+        $config = $container->get('VuFind\Config\PluginManager')
+            ->get('config');
+        $translator = $container->get('VuFind\Translator');
+        $yamlReader = $container->get('VuFind\Config\YamlReader');
+
+        return new Form($config['Feedback'] ?? null, $translator, $yamlReader);
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/FormFactory.php]: This should construct `$requestedName` rather than the hard-coded Form class in order to make it reusable when others extend the service. The change will require you to rename the service from VuFind\Form to VuFind\Form\Form -- looks a little weird, but it's our convention now to match service names to fully qualified class names.

File: module/VuFind/src/VuFind/Form.php [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+     * @var array
+     */
+    protected $messages;
+
+    /**
+     * Default form config (from config.ini > Feedback)
+     *
+     * @var array
+     */
+    protected $defaultConfig;
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Form.php]: Unless I'm missing something, this property is no longer being used -- I think it can be removed from here, the constructor, and the factory.

File: languages/en.ini [hunk 9]
```diff
@@ -382,6 +382,9 @@ fav_list_delete_fail = "Sorry, an error has occurred. Your list was not deleted.
 Favorites = "Saved Items"
 Fee = "Fee"
 Feedback = "Feedback"
+feedback_email = "Email"
```
Comment (demiankatz) [inline languages/en.ini]: Looks like a couple of these strings duplicate strings from elsewhere in the language file. I have no objection to creating feedback-specific ones to aid customization, but we might as well pre-populate all the language files with reasonable defaults by copying the existing strings. There's a command line tool `php $VUFIND_HOME/public/index.php language copystring [source] [destination]` that can automate that work. I'm happy to take care of that step if you prefer!

File: config/vufind/FeedbackForms.yaml (hunk 4 above)
Comment (samuli) [reply config/vufind/FeedbackForms.yaml]: No objection, I forgot to change this back to disabled.

File: config/vufind/FeedbackForms.yaml (hunk 3 above)
Comment (samuli) [reply config/vufind/FeedbackForms.yaml]: No objection.

Comment (samuli) [reply config/vufind/FeedbackForms.yaml]: Commented, apart from the title, which is still needed.

File: config/vufind/FeedbackForms.yaml (hunk 4 above)
Comment (samuli) [reply config/vufind/FeedbackForms.yaml]: Fixed.

File: languages/en.ini (hunk 5 above)
Comment (samuli) [reply languages/en.ini]: You're right, the subject should not be translated. Possible variables within the subject are still translated.

File: module/VuFind/config/module.config.php (hunk 6 above)
Comment (samuli) [reply module/VuFind/config/module.config.php]: It seems that this is needed for non-default forms, for example /Feedback/Form/AcquisitionRequest

File: module/VuFind/src/VuFind/Form.php (hunk 8 above)
Comment (samuli) [reply module/VuFind/src/VuFind/Form.php]: Oops, this should be $defaultFormConfig which is used.

File: module/VuFind/src/VuFind/FormFactory.php (hunk 7 above)
Comment (samuli) [reply module/VuFind/src/VuFind/FormFactory.php]: Fixed

File: languages/en.ini (hunk 9 above)
Comment (samuli) [reply languages/en.ini]: Fixed

**Comment by samuli** [comment]
@demiankatz Have you had time to look at the latest changes? Please let me know if there's still something that needs to be done.

**Comment by demiankatz** [comment]
@samuli, thanks for checking in. I've looked at the changes, and at a glance they look great, but I ran out of time before I could do more thorough testing last week, and this week I'm on vacation. I'll give this a closer look when I get back -- I'm thinking about writing a couple of unit tests as part of my review to really thoroughly cover everything.

File: module/VuFind/config/module.config.php (hunk 6 above)
Comment (demiankatz) [reply module/VuFind/config/module.config.php]: See my changes in f84ec63 -- this version seems to work for me, and it's a bit simpler. Please let me know if I've somehow broken something!

**Comment by demiankatz** [comment]
@samuli,

I've had time for a bit more review. A few notes, including one new significant issue:

1.) I've added some more missing language strings.

2.) I've done some minor edits to the comments in the YAML file, mainly adding default values. Let me know if I've written anything inaccurate.

3.) It appears that allowLocalOverride defaults to false. Is there a reason for this? I can't think of a strong reason to discourage customization. In fact, do we need this setting at all?

4.) Right now, local override appears to be at the full section level. Would there be any benefit to allowing a partial override? (I realize that this could get complicated and may not be worth the effort... but it seems unfortunate to have to copy and paste everything just to enable or disable a single field). Anyway, not a show-stopping problem, but just food for thought.

5.) There is an option in the [Captcha] section of config.ini to turn on ReCaptcha for feedback forms. This feature is broken by this PR (no ReCaptcha control displays, but the form cannot be submitted, reporting "CAPTCHA failed"). Not only should we fix this, but we might also want to consider enabling or disabling Captcha controls on a form-by-form basis. I'm not sure how difficult this will be, though -- please let me know if you need assistance with the fix.

In any case, sorry for the slow reply -- I'm having some problems with my computer this week that are slowing me down; hopefully they will be repaired before too much longer!

**Comment by samuli** [comment]
@demiankatz 

> 2.) I've done some minor edits to the comments in the YAML file, mainly adding default values. Let me know if I've written anything inaccurate.

Looks good.

> 3.) It appears that allowLocalOverride defaults to false. Is there a reason for this? I can't think of a strong reason to discourage customization. In fact, do we need this setting at all?

> 4.) Right now, local override appears to be at the full section level. Would there be any benefit to allowing a partial override? (I realize that this could get complicated and may not be worth the effort... but it seems unfortunate to have to copy and paste everything just to enable or disable a single field). Anyway, not a show-stopping problem, but just food for thought.

You're right, allowLocalOverride might be a special need. I refactored config parsing so that the merging of default and local config can be overridden by extending Form::mergeLocalConfig. This allows one to implement key-by-key based override or support allowLocalOverride. Currently it returns local config like before.

> 5.) There is an option in the [Captcha] section of config.ini to turn on ReCaptcha for feedback forms. This feature is broken by this PR (no ReCaptcha control displays, but the form cannot be submitted, reporting "CAPTCHA failed"). Not only should we fix this, but we might also want to consider enabling or disabling Captcha controls on a form-by-form basis. I'm not sure how difficult this will be, though -- please let me know if you need assistance with the fix.

Yes, totally forgot the Captcha. This is now included.

File: module/VuFind/config/module.config.php (hunk 6 above)
Comment (demiankatz) [reply module/VuFind/config/module.config.php]: ...and commit 7d58181c7914d7cd4ca32a3a035dc1b2c7237524 is even simpler yet!

**Comment by demiankatz** [comment]
@samuli, thanks once again -- we're getting very close! I made some adjustments to the Captcha configuration/documentation in 246c96e, because the way the Captcha code currently works won't allow you to turn Captcha on through FeedbackForms.yaml if the 'feedback' setting isn't already turned on in config.ini -- so now you can disable Captcha through YAML, but you still have to turn it on in config.ini. I think the documentation in the file makes this clear.

I think this could probably be merged as-is if you're in a hurry to get it into master, but if you don't mind waiting a little longer, I'd still like to take a little time to write a unit test for the new Form class to help keep our test coverage up and potentially turn up edge cases we've missed.

One observation in the meantime: in the Form class, I think we can probably simplify translation handling.  Did you consider using the TranslatorAwareTrait (https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/I18n/Translator/TranslatorAwareTrait.php) instead of constructor-injecting the translator? That would allow you to replace all the $this->translator->translate() calls with $this->translate() calls. I'm also not clear on why the parseConfig method has a $translator parameter; wouldn't it be easier to simply directly access the translator property from inside the method?

**Comment by samuli** [comment]
@demiankatz 

> Did you consider using the TranslatorAwareTrait (https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/I18n/Translator/TranslatorAwareTrait.php) instead of constructor-injecting the translator? That would allow you to replace all the $this->translator->translate() calls with $this->translate() calls. I'm also not clear on why the parseConfig method has a $translator parameter; wouldn't it be easier to simply directly access the translator property from inside the method?

Sounds good, I'll refactor it to use TranslatorAwareTrait which also fixes the issue with parseConfig.

> I think this could probably be merged as-is if you're in a hurry to get it into master, but if you don't mind waiting a little longer, I'd still like to take a little time to write a unit test for the new Form class to help keep our test coverage up and potentially turn up edge cases we've missed.

Please do. I'll make the changes above tomorrow.

**Comment by samuli** [comment]
@demiankatz 

> Sounds good, I'll refactor it to use TranslatorAwareTrait which also fixes the issue with parseConfig.

This is now done.

**Comment by demiankatz** [comment]
@samuli, I was able to write an initial unit test today. In the process, I made a couple of minor changes, mainly to make the class slightly easier to test by expecting arrays instead of specific Zend Framework classes and by making more things optional.

Please take a look at my test class and feel free to add more tests if there is anything you'd particularly like to cover. I'm sure my coverage is far from perfect, since I just wanted to get a baseline starting point in place. I also am not sure if/how I should test the input filter creation. Right now the test just checks that an input filter is created but doesn't interrogate it in any detail.

Also, one last question: I noticed that the default feedback form had some help fields defined in the yaml file, but these were not defined in the language files and did not appear to be doing anything. I commented them out for now... but is something broken? Do we need to either fix or remove this functionality?

**Comment by samuli** [comment]
@demiankatz I added some tests for input validators.

>I noticed that the default feedback form had some help fields defined in the yaml file, but these were not defined in the language files and did not appear to be doing anything. I commented them out for now... but is something broken? Do we need to either fix or remove this functionality?

Form help fields (help > pre & help > post) are optional and the text is shown if the translation key is defined. These can be left as commeted out by default. 
Form element also supports a help text that is configured with the key "help". I updated the documentation to cover this.

**Comment by demiankatz** [comment]
Thanks, @samuli, for the extra test assertions and for explaining the help behavior. One last question before merging: would you object to removing the logic to remove help text when translations are missing? Generally, VuFind displays text whether or not a translation exists, and I think having this behavior here might be likely to cause user confusion. It will also cause extra work for users that don't use multiple languages in the interface. If there is a strong use case for keeping it, then we should at least be more explicit about the behavior in the documentation.

In any case, let me know what you think about this, and once we finalize that decision, I'll merge this to master.

**Comment by samuli** [comment]
@demiankatz Help texts are now displayed regardless of whether the translations are defined or not. Now it supports both use cases.

**Comment by demiankatz** [comment]
@samuli, sorry for the slow final review; I was out of the office for a few days. This looks like it's ready to go, so I have merged it. Thanks again for all of your work on this, and for your patience with the review process!
